### ISSUE TITLE
Incorrect `__hash__` method for `GapElement`

### ISSUE THREAD

**Issue by fingolfin** [sagemath/sage#35926]
I noticed the following in `src/sage/libs/gap/element.pyx:828`:
```julia
    def __hash__(self):
        """
        Make hashable.

        EXAMPLES::

            sage: hash(libgap(123))   # random output
            163512108404620371
        """
        return hash(str(self))
```
But at the same time the [Python documentation for `__hash__` says this](https://docs.python.org/3.11/reference/datamodel.html#object.__hash__):
> The only required property is that objects which compare equal have the same hash value

But the above method violates that. E.g. there are GAP matrices which are equal but print quite differently.

A good fix is difficult. The only "easy" fix I can think of is to return a constant value. Any better fix requires a case distinction based on what kind of GAP object this is; and even then in general returning the constant is the best I can think of (e.g. if the object is a group)

**Comment by nbruin** [comment]
Sage breaks Python's hash rule on a regular basis already:
```
sage: a=7
sage: b=GF(3)(7)
sage: (a==b, hash(a)==hash(b))
(True, False)
```
In general, elements from different parents aren't really safe to use as keys in dictionaries in an intensive way for that reason. For elements from the same parent it's quite reasonable to demand that hash does work as prescribed.
For gap: perhaps gap itself has a hash? They probably have associative arrays/dictionaries/sets as well? Using that (possibly mangled a bit) might work?

**Comment by fingolfin** [comment]
No, gap has no such universal hash, its "records" (closest analog to dicts) only allow string keys; there is some limited hashing but it also depends on some partitioning of object in the spirit of "hashes are comparable only if parents match" like you mention. In many cases is just impossible to provide a proper hash, e.g. if looking at words in general finitely presented groups.

**Comment by nbruin** [comment]
Well ... if you want it to be correct then making GapElement objects unhashable would  be a possibility. I don't think it's unreasonable that opaque (to python) interface objects don't pretend to be hashable. However, hash tables are used in surprisingly many places in Python, so it may break too much.

I don't think going with a constant hash function will be reasonable: it will likely break performance to such a degree that disabling hashing is comparable in impact.

There's a chance neither "solution" (constant hash or no hash) will improve practical performance of sage over using "str" as hash. But, yeah, it's clear "str" hash is buggy.

**Comment by mantepse** [comment]
I ran into a related issue, which did cost me a few days again even though I should have known.  

Let me tell the story in the way it happened, long winded, to make it clear why these bugs are relatively hard to discover.  Here is what I had:
```
            sage: P = PolynomialSpecies(ZZ, ["X"])
            sage: n = 4
            sage: S = SymmetricGroup(n)
            sage: X = S.subgroups()
            sage: def act(pi, G):  # WARNING: returning H does not work because of equality problems
            ....:     H = S.subgroup(G.conjugate(pi).gens())
            ....:     return next(K for K in X if K == H)
            sage: P((X, act), {0: range(1, n+1)})
            4*E_4 + 4*P_4 + E_2^2 + 2*X*E_3
```
Before, I defined the action in the natural way: `act = lambda pi, G: S.subgroup(G.conjugate(pi).gens())`, and it didn't work, and I could not figure out why.

The reason is that a subroutine uses `orbit_decomposition`.  This turns `X` into a set, and then essentially uses `in` to check whether an element produced by the action is in the same orbit or not.  But with a bad hash function, we cannot put things into sets:
```
            sage: G = PermutationGroup([(1,2,3),(1,2)])
            sage: G == SymmetricGroup(3)
            True
            sage: G in set([SymmetricGroup(3)])
            False
```
Note that `PermutationGroup_generic` inherits the hash function from `category_object`, where it is defined as
```
    def __hash__(self):
        """
        A default hash based on the string representation of ``self``.

        It is cached to remain consistent throughout a session, even
        if the representation changes.

        EXAMPLES::

            sage: bla = PolynomialRing(ZZ,"x")
            sage: h1 = hash(bla)
            sage: h1  # random
            -5279516879544852222
            sage: bla.rename('toto')
            sage: h2 = hash(bla)
            sage: h2  # random
            -5279516879544852222
            sage: h1 == h2
            True
        """
        if self._hash_value == -1:
            self._hash_value = hash(repr(self))
        return self._hash_value
```
Of course, the short term fix is to implement a hash for permutation groups, but in truth I think we should stop with Verdrängung - also as a society, of course :-)
